File: uhrwerk-config/src/main/java/io/qimia/uhrwerk/config/representation/Connection.java
package io.qimia.uhrwerk.config.representation;

import io.qimia.uhrwerk.config.ConfigException;

public class Connection{
    private String name;
    private JDBC jdbc;
    private S3 s3;
    private File file;
    //private static String[] allowedFormats = new String[]{"jdbc","s3","file"};

    public Connection(){}

    public JDBC getJdbc() {
        return jdbc;
    }

    public void setJdbc(JDBC jdbc) {
        this.jdbc = jdbc;
    }

    public S3 getS3() {
        return s3;
    }

    public void setS3(S3 s3) {
        this.s3 = s3;
    }

    public File getFile() {
        return file;
    }

    public void setFile(File file) {
        this.file = file;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }



    public void validate(String path){
        path += "connection/";
        int containedFormats = 0;
        if(name == null){
            throw new ConfigException("Missing field: " + path + "name");
        }
        if(jdbc == null && s3 == null && file == null){
            throw new ConfigException("Missing connection format: choose either jdbc/s3/file under path: " + path);
        }

        if(jdbc != null && s3 != null){
            throw new ConfigException("Only one connection format at the same time: jdbc/s3/file under path: " + path);
        }
        if(jdbc != null && file != null){
            throw new ConfigException("Only one connection format at the same time: jdbc/s3/file under path: " + path);
        }
        if(s3 != null && file != null){
            throw new ConfigException("Only one connection format at the same time: jdbc/s3/file under path: " + path);
        }

    }

    @Override
    public String toString() {
        return "Source{" +
                "name='" + name + '\'' +
                ", jdbc='" + jdbc + '\'' +
                ", s3='" + s3 + '\'' +
                ", file='" + file + '\'' +
                '}';
    }

}

File: uhrwerk-config/src/main/kotlin/io/qimia/uhrwerk/config/representation/Connection.kt
package io.qimia.uhrwerk.config.representation

import com.fasterxml.jackson.annotation.JsonTypeInfo;
import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonSubTypes.Type;
import io.qimia.uhrwerk.config.builders.ConfigException

@JsonTypeInfo(
    use = JsonTypeInfo.Id.NAME,
    include = JsonTypeInfo.As.PROPERTY,
    property = "type"
)
@JsonSubTypes(
    Type(value = JDBC::class, name = "jdbc"),
    Type(value = S3::class, name = "s3"),
    Type(value = File::class, name = "file"),
    Type(value = Redshift::class, name = "redshift")
)
open class Connection(
    open var name: String? = null
) {
    open fun validate(path: String) {
        var path = path
        path += "connection/"
        if (name == null) {
            throw ConfigException("Missing field: " + path + "name")
        }
    }
}